cudaextender/tests/Test_CudaextenderEnd2End.cu

```diff
@@ -0,0 +1,126 @@
+*/
+#include "gtest/gtest.h"
+
+#include <cudaextender_file_location.hpp>
+#include <claraparabricks/genomeworks/cudaextender/utils.hpp>
+#include <claraparabricks/genomeworks/cudaextender/extender.hpp>
+#include <claraparabricks/genomeworks/utils/pinned_host_vector.hpp>
+
+namespace claraparabricks
+{
+
+namespace genomeworks
+{
+
+namespace cudaextender
+{
+
+typedef struct
+{
+    std::string query_file;
+    std::string target_file;
+    std::string seed_pairs_file;
+    std::string golden_scored_segment_pairs_file;
+    int32_t score_threshold;
+} End2EndTestParam;
+
+std::vector<End2EndTestParam> getCudaextenderEnd2EndTestCases()
+{
+    std::vector<End2EndTestParam> test_cases;
+
+    End2EndTestParam test1{};
+    test1.query_file                       = std::string(CUDAEXTENDER_DATA_DIR) + "/sample.fa";
+    test1.target_file                      = std::string(CUDAEXTENDER_DATA_DIR) + "/sample.fa";
+    test1.seed_pairs_file                  = std::string(CUDAEXTENDER_DATA_DIR) + "/sample_seed_pairs.csv";
+    test1.golden_scored_segment_pairs_file = std::string(CUDAEXTENDER_DATA_DIR) + "/sample_scored_segment_pairs.csv";
+    test1.score_threshold                  = 3000;
+    test_cases.push_back(test1);
+
+    return test_cases;
+}
+
+// Testing correctness of UngappedXDrop extension End2End with default scoring matrix
+class TestCudaextenderEnd2End : public ::testing::TestWithParam<End2EndTestParam>
+{
+public:
+    void SetUp()
+    {
+        param_ = GetParam();
+        // Define Scoring Matrix
+        const int32_t score_matrix[NUC2] = {91, -114, -31, -123, -1000, -1000, -100, -9100,
+                                            -114, 100, -125, -31, -1000, -1000, -100, -9100,
+                                            -31, -125, 100, -114, -1000, -1000, -100, -9100,
+                                            -123, -31, -114, 91, -1000, -1000, -100, -9100,
+                                            -1000, -1000, -1000, -1000, -1000, -1000, -1000, -9100,
+                                            -1000, -1000, -1000, -1000, -1000, -1000, -1000, -9100,
+                                            -100, -100, -100, -100, -1000, -1000, -100, -9100,
+                                            -9100, -9100, -9100, -9100, -9100, -9100, -9100, -9100};
+        const int32_t xdrop_threshold    = 910;
+        const int32_t device_id          = 0;
+        const bool no_entropy            = false;
+        const std::size_t max_gpu_memory = cudautils::find_largest_contiguous_device_memory_section();
+        allocator_                       = create_default_device_allocator(max_gpu_memory);
+        stream_                          = make_cuda_stream();
+        ungapped_extender_               = create_extender(score_matrix, NUC2, xdrop_threshold, no_entropy, stream_.get(), device_id, allocator_);
+    }
+
+    void TearDown()
+    {
+        ungapped_extender_.reset();
+    }
+
+    void run_test()
```

Comment:
Please remove the `run_test()` method and move this code directly into the body of `TEST_P(TestCudaextenderEnd2End, TestCorrectness)`. IIRC, you don't need to adapt anything - I think TEST_P will have access to the classes' private elements - I'm not sure though.

Reply:
I tried doing that. TEST_P doesn't have access to the private elements.

Reply:
Oh yes, the members need to be `protected` only.

Reply:
I think should be okay to make them `protected` as well

Reply:
Done.